### PR TITLE
Modify experiment code for multi-GPU training

### PR DESCRIPTION
**Pull request by spinjo**
In the end this didn't require too many changes, so I would suggest merging this into main. @vbpla what do you think?

### Changes
- If multiple GPUs are specified, we launch several processes in `run.py` with one experiment class each and communication through the `nccl` protocol
- `base_experiment.py` has access to `world_size` and `rank` -> Log and plot only on `rank==0`
- Use `DistributedSampler` in the `DataLoader` of each experiment (required some extra care for the custom `DataLoader` of the `eventgen` experiment)

### Comments
- Key concepts in `DDP`: `world_size` = #GPUs, `rank` = index of current GPU
- Almost all multi-gpu stuff is handled in `run.py` and `base_experiment.py`; the only thing one has to be careful with when adding a new experiment is to use `torch.utils.data.DistributedSampler` as an input to the `DataLoader`
- Metrics can be synchronized across GPUs with e.g. `dist.all_reduce`; currently this is only done for the `loss`; we could also do that for other metrics, but that gets ugly
- DDP complains if some parameters of the model are not used in the loss, but for GATr this is common because usually only either the scalar of the multivector outputs are used; we could remove the unused outputs in the code, but that would be super ugly; so we use `find_unused_parameters=True` when calling `torch.nn.parallel.DistributedDataParallel`, this adds some extra overhead but then we don't get errors

### CODE REVIEW

**Comment by spinjo** [comment]
Some notes on why we use `torch.nn.DistributedDataParallel` (DDP) instead of `torch.nn.DataParallel` (DP):
- In essense, DP uses a single process for the main code, and only does multi-processing for the models `forward`; DDP uses independent processes for each GPU for the whole training process - which is more flexible
- DP is very simple to implement: Just add a line `self.model.net = torch.nn.DataParallel(self.model.net)` in `experiments/experiment_base.py` in the `init_model` function
- DP splits the arguments to `model` in `dim=0`. In our setting this leads to annoying issues in the following cases: (a) sparse tensors in tagging, (b) when we have to add a pseudo-batch-dim and have shapes like (1, 512, 32, 4)
- DDP essentially runs the code on each GPU individually, which is much more flexible -> We do that

**Review by vbpla — APPROVED** [review]
